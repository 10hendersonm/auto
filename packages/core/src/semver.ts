import { VersionLabel } from './release';

enum SEMVER {
  major = 'major',
  premajor = 'premajor',
  minor = 'minor',
  preminor = 'preminor',
  patch = 'patch',
  prepatch = 'prepatch',
  noVersion = ''
}

export type IVersionLabels = Map<VersionLabel, string[]>;

export default SEMVER;

export function getHigherSemverTag(left: SEMVER, right: string): SEMVER {
  if (left === SEMVER.major || right === SEMVER.major) {
    return SEMVER.major;
  }

  if (left === SEMVER.minor || right === SEMVER.minor) {
    return SEMVER.minor;
  }

  return SEMVER.patch;
}

interface ISemVerOptions {
  onlyPublishWithReleaseLabel?: boolean;
  skipReleaseLabels?: string[];
}

export function calculateSemVerBump(
  labels: string[][],
  labelMap: IVersionLabels,
  { onlyPublishWithReleaseLabel, skipReleaseLabels = [] }: ISemVerOptions = {}
) {
  const labelSet = new Set<string>();
  const skip = labelMap.get('skip-release') || [];

  skip.forEach(skipLabel => {
    if (!skipReleaseLabels.includes(skipLabel)) {
      skipReleaseLabels.push(skipLabel);
    }
  });

  labels.map(pr => {
    pr.forEach(label => {
      const userLabel = [...labelMap.entries()].find(pair =>
        pair[1].includes(label)
      );
      labelSet.add(userLabel ? userLabel[0] : label);
    });
  });

  let skipRelease = false;
  let isPrerelease = false;

  if (labels.length > 0 && labels[0].length > 0) {
    const prereleaseLabels = labelMap.get('prerelease') || [];
    const releaseLabels = labelMap.get('release') || [];
    isPrerelease = labels[0].some(label => prereleaseLabels.includes(label));
    skipRelease = onlyPublishWithReleaseLabel
      ? !labels[0].some(label => releaseLabels.includes(label))
      : labels[0].some(label => skipReleaseLabels.includes(label));
  }

  const version = [...labelSet].reduce(getHigherSemverTag, SEMVER.patch);

  if (skipRelease) {
    return SEMVER.noVersion;
  }

  if (isPrerelease) {
    if (version === SEMVER.major) {
      return SEMVER.premajor;
    }

    if (version === SEMVER.minor) {
      return SEMVER.preminor;
    }

    return SEMVER.prepatch;
  }

  return version;
}
