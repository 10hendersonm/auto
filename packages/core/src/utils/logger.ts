import signale, { Signale } from 'signale';

export interface ILogger {
  /** The level at which to log messages */
  logLevel: LogLevel;
  /** The default logger. Always on */
  log: signale.Signale<signale.DefaultMethods>;
  /** The verbose log. Has more debug logs */
  verbose: signale.Signale<signale.DefaultMethods>;
  /** The very verbose log. Has all debug logs */
  veryVerbose: signale.Signale<signale.DefaultMethods>;
}

/** Create a dummy logger for testing. */
export function dummyLog(): ILogger {
  return {
    logLevel: undefined,
    log: new Signale({ disabled: true }),
    verbose: new Signale({ disabled: true }),
    veryVerbose: new Signale({ disabled: true })
  };
}

export type LogLevel = 'verbose' | 'veryVerbose' | undefined;

/** Create a logger the the given log level. */
export default function createLog(mode: LogLevel): ILogger {
  return {
    logLevel: mode,
    log: new Signale(),
    verbose: new Signale({
      disabled: mode !== 'verbose' && mode !== 'veryVerbose'
    }),
    veryVerbose: new Signale({
      disabled: mode !== 'veryVerbose'
    })
  };
}
