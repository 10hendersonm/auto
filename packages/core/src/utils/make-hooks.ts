import {
  AsyncParallelHook,
  AsyncSeriesBailHook,
  AsyncSeriesWaterfallHook,
  SyncHook,
  SyncWaterfallHook
} from 'tapable';
import { IAutoHooks } from '../auto';
import { IChangelogHooks } from '../changelog';
import { ILogParseHooks } from '../log-parse';
import { IReleaseHooks } from '../release';

export const makeHooks = (): IAutoHooks => ({
  beforeRun: new SyncHook(['config']),
  modifyConfig: new SyncWaterfallHook(['config']),
  beforeShipIt: new SyncHook([]),
  afterShipIt: new AsyncParallelHook(['version', 'commits']),
  afterRelease: new AsyncParallelHook(['releaseInfo']),
  onCreateRelease: new SyncHook(['options']),
  onCreateChangelog: new SyncHook(['changelog', 'version']),
  onCreateLogParse: new SyncHook(['logParse']),
  getAuthor: new AsyncSeriesBailHook([]),
  getPreviousVersion: new AsyncSeriesBailHook(['prefixRelease']),
  getRepository: new AsyncSeriesBailHook([]),
  version: new AsyncParallelHook(['version']),
  afterVersion: new AsyncParallelHook([]),
  publish: new AsyncParallelHook(['version']),
  afterPublish: new AsyncParallelHook([]),
  canary: new AsyncSeriesBailHook(['canaryVersion', 'postFix'])
});

export const makeReleaseHooks = (): IReleaseHooks => ({
  onCreateChangelog: new SyncHook(['changelog', 'version']),
  createChangelogTitle: new AsyncSeriesBailHook([]),
  onCreateLogParse: new SyncHook(['logParse'])
});

export const makeLogParseHooks = (): ILogParseHooks => ({
  parseCommit: new AsyncSeriesWaterfallHook(['commit']),
  omitCommit: new AsyncSeriesBailHook(['commit'])
});

export const makeChangelogHooks = (): IChangelogHooks => ({
  renderChangelogLine: new AsyncSeriesWaterfallHook(['lineData']),
  renderChangelogTitle: new AsyncSeriesBailHook(['commits', 'lineRender']),
  renderChangelogAuthor: new AsyncSeriesBailHook([
    'author',
    'commit',
    'options'
  ]),
  renderChangelogAuthorLine: new AsyncSeriesBailHook(['author', 'user']),
  omitReleaseNotes: new AsyncSeriesBailHook(['commit'])
});
