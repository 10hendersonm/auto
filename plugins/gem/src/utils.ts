import * as fs from "fs";
import { promisify } from "util";

export const mkdir = promisify(fs.mkdir);
export const readFile = promisify(fs.readFile);
export const writeFile = promisify(fs.writeFile);
