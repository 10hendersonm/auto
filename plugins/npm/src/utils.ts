import * as fs from "fs";
import { promisify } from "util";

export const readFile = promisify(fs.readFile);
export const writeFile = promisify(fs.writeFile);

/** Check if the project is a monorepo */
export const isMonorepo = () => fs.existsSync("lerna.json");

/** Parse the lerna.json file. */
export const getLernaJson = () => {
  try {
    return JSON.parse(fs.readFileSync("lerna.json", "utf8"));
  } catch (error) {
    return {};
  }
};
