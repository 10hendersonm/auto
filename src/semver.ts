import { VersionLabel } from './github-release';

enum SEMVER {
  major = 'major',
  premajor = 'premajor',
  minor = 'minor',
  preminor = 'preminor',
  patch = 'patch',
  prepatch = 'prepatch',
  noVersion = ''
}

export type IVersionLabels = Map<VersionLabel, string>;

export default SEMVER;

export function getHigherSemverTag(left: string, right: string): SEMVER {
  if (left === SEMVER.major || right === SEMVER.major) {
    return SEMVER.major;
  }

  if (left === SEMVER.minor || right === SEMVER.minor) {
    return SEMVER.minor;
  }

  return SEMVER.patch;
}

interface ISemVerOptions {
  onlyPublishWithReleaseLabel?: boolean;
  skipReleaseLabels?: string[];
}

export function calculateSemVerBump(
  labels: string[][],
  labelMap: IVersionLabels,
  { onlyPublishWithReleaseLabel, skipReleaseLabels = [] }: ISemVerOptions = {}
) {
  const labelSet = new Set<string>();

  if (!skipReleaseLabels.includes(labelMap.get('skip-release')!)) {
    skipReleaseLabels.push(labelMap.get('skip-release')!);
  }

  labels.map(pr =>
    pr.forEach(label => {
      const userLabel = [...labelMap.entries()].find(pair => pair[1] === label);
      labelSet.add(userLabel ? userLabel[0] : label);
    })
  );

  let skipRelease = false;
  let isPrerelease = false;

  if (labels.length > 0 && labels[0].length > 0) {
    isPrerelease = labels[0].includes(labelMap.get('prerelease')!);
    skipRelease = onlyPublishWithReleaseLabel
      ? !labels[0].includes(labelMap.get('release')!)
      : !!labels[0].find(label => skipReleaseLabels.includes(label));
  }

  const version = [...labelSet].reduce(
    (semver: SEMVER, nextLabel) => getHigherSemverTag(semver, nextLabel),
    SEMVER.patch
  );

  if (skipRelease) {
    return SEMVER.noVersion;
  }

  if (isPrerelease) {
    if (version === SEMVER.major) {
      return SEMVER.premajor;
    }

    if (version === SEMVER.minor) {
      return SEMVER.preminor;
    }

    return SEMVER.prepatch;
  }

  return version;
}
