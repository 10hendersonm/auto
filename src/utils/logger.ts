import signale, { Signale } from 'signale';

export interface ILogger {
  log: signale.Signale<signale.DefaultMethods>;
  verbose: signale.Signale<signale.DefaultMethods>;
  veryVerbose: signale.Signale<signale.DefaultMethods>;
}

export function dummyLog(): ILogger {
  return {
    log: new Signale({ disabled: true }),
    verbose: new Signale({ disabled: true }),
    veryVerbose: new Signale({ disabled: true })
  };
}

export type LogLevel = 'verbose' | 'veryVerbose' | undefined;

export default function createLog(mode: LogLevel): ILogger {
  return {
    log: new Signale(),
    verbose: new Signale({
      disabled: mode !== 'verbose' && mode !== 'veryVerbose'
    }),
    veryVerbose: new Signale({
      disabled: mode !== 'veryVerbose'
    })
  };
}
