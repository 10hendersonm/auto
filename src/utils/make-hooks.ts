import { AsyncSeriesBailHook, AsyncSeriesHook, SyncHook } from 'tapable';
import { IGitHubReleaseHooks } from '../github-release';
import { IChangelogHooks } from '../log-parse';
import { IAutoHooks } from '../main';

export const makeHooks = (): IAutoHooks => ({
  beforeRun: new SyncHook(['config']),
  beforeShipIt: new SyncHook([]),
  onCreateGitHubRelease: new SyncHook(['config']),
  onCreateChangelog: new SyncHook(['gitHubReleaseConfig']),
  getAuthor: new AsyncSeriesBailHook([]),
  getPreviousVersion: new AsyncSeriesBailHook(['prefixRelease']),
  getRepository: new AsyncSeriesBailHook([]),
  publish: new AsyncSeriesHook(['version'])
});

export const makeGitHubReleaseHooks = (): IGitHubReleaseHooks => ({
  onCreateChangelog: new SyncHook(['changelog'])
});

export const makeChangelogHooks = (): IChangelogHooks => ({
  renderChangelogLine: new AsyncSeriesBailHook(['commits', 'lineRender']),
  renderChangelogTitle: new AsyncSeriesBailHook(['commits', 'lineRender']),
  renderChangelogAuthor: new AsyncSeriesBailHook([
    'author',
    'commit',
    'options'
  ]),
  renderChangelogAuthorLine: new AsyncSeriesBailHook(['author', 'user'])
});
