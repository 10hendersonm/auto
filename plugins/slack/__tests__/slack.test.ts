import Auto from "@auto-it/core";
import makeCommitFromMsg from "@auto-it/core/src/__tests__/make-commit-from-msg";
import { dummyLog } from "@auto-it/core/src/utils/logger";
import { makeHooks } from "@auto-it/core/src/utils/make-hooks";
import { defaultLabels } from "@auto-it/core/dist/semver";
import { execSync } from "child_process";
import createHttpsProxyAgent from "https-proxy-agent";

import SlackPlugin, { sanitizeMarkdown, convertToBlocks } from "../src";
import endent from "endent";

const fetchSpy = jest.fn();
// @ts-ignore
jest.mock("node-fetch", () => (...args) => {
  fetchSpy(...args);
});

beforeEach(() => {
  fetchSpy.mockClear();
});

const mockResponse = [
  {
    data: {
      html_url: "https://git.hub/some/project/releases/v1.0.0",
      name: "v1.0.0",
    },
  },
];

// For the purpose of this test, we use the current branch as the "prerelease" branch to fake being on a "next" branch
const nextBranch = execSync("git rev-parse --abbrev-ref HEAD", {
  encoding: "utf8",
}).trim();

const mockAuto = {
  git: {},
  logger: dummyLog(),
} as any;

describe("postToSlack", () => {
  test("doesn't post with no new version", async () => {
    const plugin = new SlackPlugin("https://custom-slack-url");
    const hooks = makeHooks();

    jest.spyOn(plugin, "createPost").mockImplementation();
    // @ts-ignore
    plugin.apply({ hooks } as Auto);

    await hooks.afterRelease.promise({
      lastRelease: "0.1.0",
      commits: [],
      releaseNotes: "# My Notes",
    });

    expect(plugin.createPost).not.toHaveBeenCalled();
  });

  test("doesn't post with no commits", async () => {
    const plugin = new SlackPlugin("https://custom-slack-url");
    const hooks = makeHooks();

    jest.spyOn(plugin, "createPost").mockImplementation();
    // @ts-ignore
    plugin.apply({ hooks, options: {} } as Auto);

    await hooks.afterRelease.promise({
      newVersion: "1.0.0",
      lastRelease: "0.1.0",
      commits: [],
      releaseNotes: "# My Notes",
    });

    expect(plugin.createPost).not.toHaveBeenCalled();
  });

  test("doesn't post with skip release label", async () => {
    const plugin = new SlackPlugin("https://custom-slack-url");
    const hooks = makeHooks();

    jest.spyOn(plugin, "createPost").mockImplementation();
    // @ts-ignore
    plugin.apply({
      hooks,
      options: {},
      config: { labels: defaultLabels },
    } as Auto);

    await hooks.afterRelease.promise({
      newVersion: "1.0.0",
      lastRelease: "0.1.0",
      commits: [makeCommitFromMsg("skipped", { labels: ["skip-release"] })],
      releaseNotes: "# My Notes",
    });

    expect(plugin.createPost).not.toHaveBeenCalled();
  });

  test("doesn't post without url", async () => {
    // @ts-ignore
    const plugin = new SlackPlugin({ url: undefined });
    const hooks = makeHooks();

    jest.spyOn(plugin, "createPost").mockImplementation();
    // @ts-ignore
    plugin.apply({ hooks, options: {} } as Auto);

    await expect(
      hooks.afterRelease.promise({
        newVersion: "1.0.0",
        lastRelease: "0.1.0",
        commits: [makeCommitFromMsg("a patch")],
        releaseNotes: "# My Notes",
      })
    ).rejects.toBeInstanceOf(Error);
  });

  test("doesn't post when prerelease branch and using default prereleasePublish setting", async () => {
    // @ts-ignore
    const plugin = new SlackPlugin({
      url: "https://custom-slack-url",
    });
    const hooks = makeHooks();

    jest.spyOn(plugin, "createPost").mockImplementation();
    // @ts-ignore
    plugin.apply({
      ...mockAuto,
      hooks,
      options: {},
      config: {
        prereleaseBranches: [nextBranch],
        labels: defaultLabels,
      },
    } as Auto);

    await hooks.afterRelease.promise({
      newVersion: "1.0.0",
      lastRelease: "0.1.0",
      commits: [makeCommitFromMsg("a patch")],
      releaseNotes: "# My Notes",
    });
    expect(fetchSpy).not.toHaveBeenCalled();
  });

  test("doesn't post when prerelease branch setting is false", async () => {
    // @ts-ignore
    const plugin = new SlackPlugin({
      url: "https://custom-slack-url",
      publishPreRelease: false,
    });
    const hooks = makeHooks();

    jest.spyOn(plugin, "createPost").mockImplementation();
    // @ts-ignore
    plugin.apply({
      ...mockAuto,
      hooks,
      options: {},
      config: {
        prereleaseBranches: [nextBranch],
        labels: defaultLabels,
      },
    } as Auto);

    await hooks.afterRelease.promise({
      newVersion: "1.0.0",
      lastRelease: "0.1.0",
      commits: [makeCommitFromMsg("a patch")],
      releaseNotes: "# My Notes",
    });
    expect(fetchSpy).not.toHaveBeenCalled();
  });

  test("posts when prerelease branch setting is true", async () => {
    // @ts-ignore
    const plugin = new SlackPlugin({
      url: "https://custom-slack-url",
      publishPreRelease: true,
    });
    const hooks = makeHooks();

    jest.spyOn(plugin, "createPost").mockImplementation();
    // @ts-ignore
    plugin.apply({
      ...mockAuto,
      hooks,
      options: {},
      config: { prereleaseBranches: ["next"], labels: defaultLabels },
    } as Auto);

    await hooks.afterRelease.promise({
      newVersion: "1.0.0",
      lastRelease: "0.1.0",
      commits: [makeCommitFromMsg("a patch")],
      releaseNotes: "# My Notes",
      // @ts-ignore
      response: mockResponse,
    });
    expect(plugin.createPost).toHaveBeenCalledTimes(1);
  });

  test("should warn when no token", async () => {
    const plugin = new SlackPlugin("https://custom-slack-url");
    const logger = dummyLog();
    jest.spyOn(logger.verbose, "warn").mockImplementation();
    process.env.SLACK_TOKEN = "";

    await plugin.createPost(
      { ...mockAuto, logger } as Auto,
      "New Releases: 1.0.0",
      sanitizeMarkdown("# My Notes\n- PR [some link](google.com)"),
      [{ data: { tag_name: "1.0.0", html_url: "https://google.com" } }] as any,
      undefined
    );

    expect(logger.verbose.warn).toHaveBeenCalled();
  });

  test("should call slack api with minimal config", async () => {
    const plugin = new SlackPlugin("https://custom-slack-url");
    process.env.SLACK_TOKEN = "MY_TOKEN";

    await plugin.createPost(
      mockAuto,
      "New Releases: 1.0.0",
      sanitizeMarkdown("# My Notes\n- PR [some link](google.com)"),
      [{ data: { tag_name: "1.0.0", html_url: "https://google.com" } }] as any,
      undefined
    );

    expect(fetchSpy).toHaveBeenCalled();
    expect(fetchSpy.mock.calls[0][0]).toBe(
      "https://custom-slack-url?token=MY_TOKEN"
    );
    expect(fetchSpy.mock.calls[0][1].agent).toBeUndefined();
    expect(JSON.parse(fetchSpy.mock.calls[0][1].body)).toMatchSnapshot();
  });

  test("should add more indents to nested lists", async () => {
    const plugin = new SlackPlugin("https://custom-slack-url");
    process.env.SLACK_TOKEN = "MY_TOKEN";

    await plugin.createPost(
      mockAuto,
      "New Releases: 1.0.0",
      sanitizeMarkdown(
        "# My Notes\n- PR [some link](google.com)\n - Another note"
      ),
      [{ data: { tag_name: "1.0.0", html_url: "https://google.com" } }] as any,
      undefined
    );

    expect(fetchSpy).toHaveBeenCalled();
    expect(fetchSpy.mock.calls[0][0]).toBe(
      "https://custom-slack-url?token=MY_TOKEN"
    );
    expect(fetchSpy.mock.calls[0][1].agent).toBeUndefined();
    expect(JSON.parse(fetchSpy.mock.calls[0][1].body)).toMatchSnapshot();
  });

  test("should add more indents to nested lists - 2 spaces", async () => {
    const plugin = new SlackPlugin("https://custom-slack-url");
    process.env.SLACK_TOKEN = "MY_TOKEN";

    await plugin.createPost(
      mockAuto,
      "New Releases: 1.0.0",
      sanitizeMarkdown(
        "# My Notes\n- PR [some link](google.com)\n  - Another note"
      ),
      [{ data: { tag_name: "1.0.0", html_url: "https://google.com" } }] as any,
      undefined
    );

    expect(fetchSpy).toHaveBeenCalled();
    expect(fetchSpy.mock.calls[0][0]).toBe(
      "https://custom-slack-url?token=MY_TOKEN"
    );
    expect(fetchSpy.mock.calls[0][1].agent).toBeUndefined();
    expect(JSON.parse(fetchSpy.mock.calls[0][1].body)).toMatchSnapshot();
  });

  test("should call slack api through http proxy", async () => {
    const plugin = new SlackPlugin("https://custom-slack-url");
    process.env.SLACK_TOKEN = "MY_TOKEN";
    process.env.http_proxy = "http-proxy";

    await plugin.createPost(
      mockAuto,
      "New Releases: 1.0.0",
      sanitizeMarkdown("# My Notes\n- PR [some link](google.com)"),
      [{ data: { tag_name: "1.0.0", html_url: "https://google.com" } }] as any,
      createHttpsProxyAgent("mock-url")
    );

    expect(fetchSpy).toHaveBeenCalled();
    expect(fetchSpy.mock.calls[0][0]).toBe(
      "https://custom-slack-url?token=MY_TOKEN"
    );
    expect(fetchSpy.mock.calls[0][1].agent).not.toBeUndefined();
    expect(JSON.parse(fetchSpy.mock.calls[0][1].body)).toMatchSnapshot();
  });

  test("should remove markdown code types from block", async () => {
    const plugin = new SlackPlugin("https://custom-slack-url");
    process.env.SLACK_TOKEN = "MY_TOKEN";

    await plugin.createPost(
      mockAuto,
      "New Releases: 1.0.0",
      sanitizeMarkdown(
        `# My Notes\n\`\`\`json\n{ "foo": "bar" }\`\`\`\n- PR [some link](google.com)`
      ),
      [{ data: { tag_name: "1.0.0", html_url: "https://google.com" } }] as any,
      undefined
    );

    expect(JSON.parse(fetchSpy.mock.calls[0][1].body)).toMatchSnapshot();
  });

  test("should call slack api through https proxy", async () => {
    const plugin = new SlackPlugin("https://custom-slack-url");
    process.env.SLACK_TOKEN = "MY_TOKEN";
    process.env.https_proxy = "https-proxy";

    await plugin.createPost(
      mockAuto,
      "New Releases: 1.0.0",
      sanitizeMarkdown("# My Notes\n- PR [some link](google.com)"),
      [{ data: { tag_name: "1.0.0", html_url: "https://google.com" } }] as any,
      createHttpsProxyAgent("mock-url")
    );

    expect(fetchSpy).toHaveBeenCalled();
    expect(fetchSpy.mock.calls[0][0]).toBe(
      "https://custom-slack-url?token=MY_TOKEN"
    );
    expect(fetchSpy.mock.calls[0][1].agent).not.toBeUndefined();
    expect(JSON.parse(fetchSpy.mock.calls[0][1].body)).toMatchSnapshot();
  });

  test("should call slack api", async () => {
    const plugin = new SlackPlugin({ url: "https://custom-slack-url" });
    const hooks = makeHooks();
    process.env.SLACK_TOKEN = "MY_TOKEN";
    plugin.apply({ hooks, options: {}, ...mockAuto } as Auto);

    await hooks.afterRelease.promise({
      newVersion: "1.0.0",
      lastRelease: "0.1.0",
      commits: [makeCommitFromMsg("a patch")],
      releaseNotes: "# My Notes\n- PR [some link](google.com)",
      // @ts-ignore
      response: mockResponse,
    });

    expect(fetchSpy).toHaveBeenCalled();
    expect(fetchSpy.mock.calls[0][0]).toBe(
      "https://custom-slack-url?token=MY_TOKEN"
    );
    expect(JSON.parse(fetchSpy.mock.calls[0][1].body)).toMatchSnapshot();
  });

  test("should call slack api with custom atTarget", async () => {
    const plugin = new SlackPlugin({
      url: "https://custom-slack-url",
      atTarget: "here",
    });
    const hooks = makeHooks();
    process.env.SLACK_TOKEN = "MY_TOKEN";
    plugin.apply({ hooks, options: {}, ...mockAuto } as Auto);

    await hooks.afterRelease.promise({
      newVersion: "1.0.0",
      lastRelease: "0.1.0",
      commits: [makeCommitFromMsg("a patch")],
      releaseNotes: "# My Notes\n- PR [some link](google.com)",
      // @ts-ignore
      response: mockResponse,
    });

    expect(fetchSpy).toHaveBeenCalled();
    expect(fetchSpy.mock.calls[0][0]).toBe(
      "https://custom-slack-url?token=MY_TOKEN"
    );
    expect(fetchSpy.mock.calls[0][1].body.includes("@here")).toBe(true);
    expect(JSON.parse(fetchSpy.mock.calls[0][1].body)).toMatchSnapshot();
  });

  test("should call slack api in env var", async () => {
    process.env.SLACK_WEBHOOK_URL = "https://foo.bar";
    const plugin = new SlackPlugin();
    const hooks = makeHooks();
    process.env.SLACK_TOKEN = "MY_TOKEN";
    plugin.apply({ hooks, options: {}, ...mockAuto } as Auto);

    await hooks.afterRelease.promise({
      newVersion: "1.0.0",
      lastRelease: "0.1.0",
      commits: [makeCommitFromMsg("a patch")],
      releaseNotes: "# My Notes\n- PR [some link](google.com)",
      // @ts-ignore
      response: mockResponse,
    });

    expect(fetchSpy).toHaveBeenCalled();
    expect(fetchSpy.mock.calls[0][0]).toBe("https://foo.bar?token=MY_TOKEN");
    expect(JSON.parse(fetchSpy.mock.calls[0][1].body)).toMatchSnapshot();
  });

  test("should add title", async () => {
    process.env.SLACK_WEBHOOK_URL = "https://foo.bar";
    const plugin = new SlackPlugin({ title: "My Cool Project" });
    const hooks = makeHooks();
    process.env.SLACK_TOKEN = "MY_TOKEN";
    plugin.apply({ hooks, options: {}, ...mockAuto } as Auto);

    await hooks.afterRelease.promise({
      newVersion: "1.0.0",
      lastRelease: "0.1.0",
      commits: [makeCommitFromMsg("a patch")],
      releaseNotes: "# My Notes\n- PR [some link](google.com)",
      // @ts-ignore
      response: mockResponse,
    });

    expect(fetchSpy).toHaveBeenCalled();
    expect(JSON.parse(fetchSpy.mock.calls[0][1].body)).toMatchSnapshot();
  });
});

describe("convertToBlocks", () => {
  test("work for simple case", () => {
    expect(
      convertToBlocks(
        sanitizeMarkdown(endent`
          #### 🐛 Bug Fix

          - build the s3 plugin [#1804](https://github.com/intuit/auto/pull/1804) ([@hipstersmoothie](https://github.com/hipstersmoothie))
          
          #### Authors: 1
          
          - Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))    
      `)
      )
    ).toMatchSnapshot();
  });

  test("work for simple additional release notes", () => {
    expect(
      convertToBlocks(
        sanitizeMarkdown(endent`
          ### Release Notes

          #### Don't create "Canary Release Assets" during non-canary builds + Change that releases tag to valid semver ([#1802](https://github.com/intuit/auto/pull/1802))
          
          This release changes the tag used for the "Canary Releases Assets" created by the \`upload-assets\` plugin to be \`0.0.0-canary\`.
          This new tag is a valid semantic version and can be used with other auto commands.
          
          ---
          
          #### 🐛 Bug Fix
          
          - \`@auto-it/upload-assets\`
            - Don't create "Canary Release Assets" during non-canary builds + Change that releases tag to valid semver [#1802](https://github.com/intuit/auto/pull/1802) ([@hipstersmoothie](https://github.com/hipstersmoothie))
          
          #### 📝 Documentation
          
          - add automated TOC to hooks documentation [#1801](https://github.com/intuit/auto/pull/1801) ([@hipstersmoothie](https://github.com/hipstersmoothie))
          
          #### Authors: 1
          
          - Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))
      `)
      )
    ).toMatchSnapshot();
  });

  test("work for simple additional release notes with code examples", () => {
    expect(
      convertToBlocks(
        sanitizeMarkdown(endent`
          ### Release Notes

          #### Don't create "Canary Release Assets" during non-canary builds + Change that releases tag to valid semver ([#1802](https://github.com/intuit/auto/pull/1802))
          
          \`\`\`md
          > Code example
          \`\`\`
          
          ---
          
          #### 🐛 Bug Fix
          
          - \`@auto-it/upload-assets\`
            - Don't create "Canary Release Assets" during non-canary builds + Change that releases tag to valid semver [#1802](https://github.com/intuit/auto/pull/1802) ([@hipstersmoothie](https://github.com/hipstersmoothie))
          
          #### 📝 Documentation
          
          - add automated TOC to hooks documentation [#1801](https://github.com/intuit/auto/pull/1801) ([@hipstersmoothie](https://github.com/hipstersmoothie))
          
          #### Authors: 1
          
          - Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))
      `)
      )
    ).toMatchSnapshot();
  });
});
