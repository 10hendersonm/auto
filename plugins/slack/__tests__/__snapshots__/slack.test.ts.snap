// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postToSlack should add more indents to nested lists - 2 spaces 1`] = `"{\\"text\\":\\"@channel: New release *<https://git.hub/some/project/releases/v1.0.0|v1.0.0>*\\\\n*My Notes*\\\\n• PR <google.com|some link>\\\\n   • Another note\\",\\"link_names\\":1}"`;

exports[`postToSlack should add more indents to nested lists 1`] = `"{\\"text\\":\\"@channel: New release *<https://git.hub/some/project/releases/v1.0.0|v1.0.0>*\\\\n*My Notes*\\\\n• PR <google.com|some link>\\\\n   • Another note\\",\\"link_names\\":1}"`;

exports[`postToSlack should add title 1`] = `"{\\"text\\":\\"My Cool Project\\\\n\\\\n@channel: New release *<https://git.hub/some/project/releases/v1.0.0|v1.0.0>*\\\\n*My Notes*\\\\n• PR <google.com|some link>\\",\\"link_names\\":1}"`;

exports[`postToSlack should call slack api 1`] = `"{\\"text\\":\\"@channel: New release *<https://git.hub/some/project/releases/v1.0.0|v1.0.0>*\\\\n*My Notes*\\\\n• PR <google.com|some link>\\",\\"link_names\\":1}"`;

exports[`postToSlack should call slack api in env var 1`] = `"{\\"text\\":\\"@channel: New release *<https://git.hub/some/project/releases/v1.0.0|v1.0.0>*\\\\n*My Notes*\\\\n• PR <google.com|some link>\\",\\"link_names\\":1}"`;

exports[`postToSlack should call slack api through http proxy 1`] = `"{\\"text\\":\\"@channel: New release *<https://git.hub/some/project/releases/v1.0.0|v1.0.0>*\\\\n*My Notes*\\\\n• PR <google.com|some link>\\",\\"link_names\\":1}"`;

exports[`postToSlack should call slack api through https proxy 1`] = `"{\\"text\\":\\"@channel: New release *<https://git.hub/some/project/releases/v1.0.0|v1.0.0>*\\\\n*My Notes*\\\\n• PR <google.com|some link>\\",\\"link_names\\":1}"`;

exports[`postToSlack should call slack api with custom atTarget 1`] = `"{\\"text\\":\\"@here: New release *<https://git.hub/some/project/releases/v1.0.0|v1.0.0>*\\\\n*My Notes*\\\\n• PR <google.com|some link>\\",\\"link_names\\":1}"`;

exports[`postToSlack should call slack api with minimal config 1`] = `"{\\"text\\":\\"@channel: New release *<https://git.hub/some/project/releases/v1.0.0|v1.0.0>*\\\\n*My Notes*\\\\n• PR <google.com|some link>\\",\\"link_names\\":1}"`;

exports[`postToSlack should remove markdown code types from block 1`] = `"{\\"text\\":\\"@channel: New release *<https://git.hub/some/project/releases/v1.0.0|v1.0.0>*\\\\n*My Notes*\\\\n\`json\`:\\\\n\\\\n\`\`\`\\\\n{ \\\\\\"foo\\\\\\": \\\\\\"bar\\\\\\" }\`\`\`\\\\n• PR <google.com|some link>\\",\\"link_names\\":1}"`;
